### PR TITLE
Dirty LocalStorage monkeypatch for error catching

### PR DESCRIPTION
**Pull request by Palid**
WARNING: Dirty hack to be removed later in the future, via a refactor
to a proper storage-service object instead of monkeypatching all of the
localStorage calls that may or may not throw an error.
This should be properly handled in-place whenever it can throw, but this
has to do for now.

Add a way to listen to any kinds of errors thrown by LocalStorage
methods so that we can handle them on the top level and at least show
some details to the user that the app has broken due to LocalStorage
failing hard.

Related: https://github.com/vector-im/element-web/issues/18423

Requires, in following order:
* https://github.com/matrix-org/matrix-js-sdk/pull/2018
* https://github.com/matrix-org/matrix-js-sdk/pull/2019

### CODE REVIEW

**Review by gsouquet — APPROVED** [review]
Looks good to me!
I'd be interested in hearing a bit more about the final solution we're aiming for and why we have not decided to head that way right now?

What are the risks that you foresee with this implementation that the other solution would solve?

**Review by gsouquet — DISMISSED** [review]
Hit the wrong review button, I believe we need to change the `window.localStorage` before this can be merged

File: src/vector/localstorage-fix.ts [hunk 1]
```diff
@@ -1,12 +1,89 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import {
+    localStorageErrorsEventsEmitter,
+    LocalStorageErrors,
+} from 'matrix-js-sdk/src/store/local-storage-events-emitter';
+
+const ls = window.localStorage;
```
[the remark this answers is not in the document]
Comment (Palid) [reply src/vector/localstorage-fix.ts:22]: It is very much invalid though, there's no localStorage in nodejs. We should probably fix this issue too.

[the remark this answers is not in the document]
Comment (Palid) [reply src/vector/localstorage-fix.ts:22]: It's only `element-web` change too, so it doesn't matter if I use `global` or `window` here.

File: src/vector/localstorage-fix.ts [hunk 2]
```diff
@@ -1,12 +1,89 @@
...
+            localStorageErrorsEventsEmitter.emit(LocalStorageErrors.RemoveItemError, e);
+            if (isQuotaExceededError(e)) {
+                localStorageErrorsEventsEmitter.emit(LocalStorageErrors.QuotaExceededError, e);
+            }
+            throw e;
+        }
+    };
+    ls.clear = () => {
+        return clear.call(ls);
+    };
```
[the remark this answers is not in the document]
Comment (Palid) [reply src/vector/localstorage-fix.ts:88]: Removing

**Comment by Palid** [comment]
> Hit the wrong review button, I believe we need to change the `window.localStorage` before this can be merged

Would've been an issue in `matrix-js-sdk`, not in `element-web`.

**Comment by Palid** [comment]
@gsouquet this solution is pretty much the lowest amount of work that __kind of__ gives us what we need - error handling in such a way that we can tell user that something's wrong with their disk and the app couldn't save the data properly. 
It's very much a crap implementation, as the proper one would require going through all the usages, making a proper data-handling service and adding a proper way to propagate errors from this service and handling them in call places, so that we'd know what actually did fail.
In here, the only thing we can get is that "something has failed" or "something has HORRIBLY failed".
Good enough, ship it, but there's a lot of room for improvement.

**Comment by novocaine** [comment]
> @gsouquet this solution is pretty much the lowest amount of work that **kind of** gives us what we need - error handling in such a way that we can tell user that something's wrong with their disk and the app couldn't save the data properly. It's very much a crap implementation, as the proper one would require going through all the usages, making a proper data-handling service and adding a proper way to propagate errors from this service and handling them in call places, so that we'd know what actually did fail. In here, the only thing we can get is that "something has failed" or "something has HORRIBLY failed". Good enough, ship it, but there's a lot of room for improvement.

I'm all for incremental solutions, but I'm a little bit uncertain about jumping straight to something flagged quite so loudly as "crap", especially given we're looking for ways to introduce less technical debt. Maybe it would be more effective to explore a little bit further into a good solution and assess how much effort that would take before committing to this one?

Also bearing in mind my acceptance criteria for https://github.com/vector-im/element-web/issues/18423 is 

> The app should gracefully handle localStorage write failure. It should inform the user and cleanly abort the current operation.

I don't think this approach can satisfy the second part of that?

**Comment by Palid** [comment]
@novocaine informing user and cleanly aborting current operation can be done within this solution, as we can have access to the key/value within the propagated error event. In most cases it'll be out of storage/save error/disk access error, so if any of those happens we can't really do much, as "backup saving" more can cause even more data corruption.

Currently there's 83uses of `localStorage.setItem`, with no real validation or error handling. If we want to do it properly and go one-by-one we need to add a lot of additional error handling.
Sure, we can group settings in one modal saying "hey, something has failed while updating your preferences", but there are also things like saving credentials (only if indexeddb fails for some reason), home server and other things, which when corrupted/failed to access/failed to save for some reason pretty much breaks the app entirely.

I very much prefer to call this approach a hack and then go with one-by-one approach, so that we handle the problem anyhow now, and then incrementally improve upon it instead of spending much more time handling every single case individually now, especially considering that we already had cases where the app has just died and we couldn't show any way to bail out to the user.

I have to add that this solution __does not__ close the ticket, it's one of the parts necessary for better user experience for now. It can be properly closed after we implement the proper error handling in every single case, but, in my opinion, having a global error handler here that basically tells you that app's dead and your HDD/SSD is probably dying too is much better than silently erroring out of existence and possibly breaking the app, leaving the user dumbfounded.

TL;DR: Hack now to make it work, proper solution that does not interfere with the hack later.

**Comment by dbkr** [comment]
My main question is what's driving this short-term solution? As far as I knew, this started breaking things when we inadvertently wrote one item to localstorage for each room the user was in, which caused it to fill up, but with that fixed, I don't know that it's causing actual breakage right now other than in the case of complete diskfill. If this were a band-aid to fix problems users were frequently experiencing in the wild, I think it would be defensible.

It's very likely that, of the 83 localstorage write, most happen very infrequently, so we could start by explicitly handling write failures on the most common and/or most important ones.

This also swallows the exception, so whilst we can display the error to user user, whatever was trying to write to localstorage will carry on as if everything had worked. Wouldn't this hinder our ability to add explicit error handling later?

Also, if we're concerned about the effort of changing all the calls to use a wrapper, remember that we pass a localstorage object into js-sdk to use, so we could just pass in our wrapper instead.

**Comment by novocaine** [comment]
FWIW, I don't think wrapping 83 calls in a function that try/catches and shows an error modal if something went wrong will take a prohibitively long amount of time. It it takes one minute to replace each direct localstorage invocation with a call to this function, this is an hour and a half of drudge work. You might be able to work out a regex to automate the refactor in less time.

Ensuring the error cleanly aborts the operation will as you say take longer, but simply wrapping it is an equivalent change to what this PR is doing, except its something we'd build upon rather than rip out later.

**Comment by novocaine** [comment]
> @novocaine I don't think it's as simple as that, we need to handle and rollback every single case based upon if it's actually a very important one or not.

This PR doesn't do that though, right? I'm proposing a like-for-like for this exact PR that we won't rip out later..

File: src/vector/localstorage-fix.ts [hunk 3]
```diff
@@ -1,12 +1,86 @@
-/**
- * Because we've been saving a lot of additional logger data in the localStorage for no particular reason
- * we need to, hopefully, unbrick user's devices by geting rid of unnecessary data.
- * */
+/*
+Copyright 2021 The Matrix.org Foundation C.I.C.
```
Comment (turt2live) [inline src/vector/localstorage-fix.ts:2]: incorrect copyright

**Review by turt2live — CHANGES_REQUESTED** [review]
I share a lot of the concerns raised in the comments already. A possible, less invasive, change might be to introduce a migration system with a numeric counter, converting thousands of localstorage records to just one.

This would also help us implement future transitions such as ways to migrate settings, build a whole new settings system, and generally make settings more rational to store data (you may notice a theme here).

**Comment by novocaine** [comment]
Closing this one as we intend to address the issue a different way
